utils/ffmpegvideodecode/ffmpeg_video_dec.cpp

```diff
@@ -0,0 +1,866 @@
+#endif
+    av_pkt_cnt_ = (av_pkt_cnt_ + 1) % av_packets_.size();
+    if (!av_pkt->data || !av_pkt->size) {
+        end_of_stream_ = true;
+    }
+
+    return 1;
+}
+
+/**
+ * @brief function to handle display picture
+ * 
+ * @param pDispInfo 
+ * @return int 0:fail 1: success
+ */
+int FFMpegVideoDecoder::HandlePictureDisplay(RocdecParserDispInfo *pDispInfo) {
+    // check if we reached eos here. This is a hack since HandlePictureDecode won't be called at the end_of_sequece. 
+    // so we need to flush FFMpeg decoder when we have received the lastpacket with 0 bytes
+    if (!last_packet_.payload_size && !end_of_stream_) {
+        AVPacket pkt = { 0 };
+    #if NO_DECODE_THREAD
+        // for testing Decoding without threading
+        DecodeAvFrame(&pkt, dec_frames_[av_frame_cnt_]);
+    #else
+        //push packet into packet q for decoding
+        PushPacket(&pkt);
+    #endif
+    }
+
+    if (b_extract_sei_message_) {
+        if (sei_message_display_q_[pDispInfo->picture_index].sei_data) {
+            // Write SEI Message
+            uint8_t *sei_buffer = (uint8_t *)(sei_message_display_q_[pDispInfo->picture_index].sei_data);
+            uint32_t sei_num_messages = sei_message_display_q_[pDispInfo->picture_index].sei_message_count;
+            RocdecSeiMessage *sei_message = sei_message_display_q_[pDispInfo->picture_index].sei_message;
+            if (fp_sei_) {
+                for (uint32_t i = 0; i < sei_num_messages; i++) {
+                    if (codec_id_ == rocDecVideoCodec_AVC || rocDecVideoCodec_HEVC) {
+                        switch (sei_message[i].sei_message_type) {
+                            case SEI_TYPE_TIME_CODE: {
+                                //todo:: check if we need to write timecode
+                            }
+                            break;
+                            case SEI_TYPE_USER_DATA_UNREGISTERED: {
+                                fwrite(sei_buffer, sei_message[i].sei_message_size, 1, fp_sei_);
+                            }
+                            break;
+                        }
+                    }
+                    if (codec_id_ == rocDecVideoCodec_AV1) {
+                        fwrite(sei_buffer, sei_message[i].sei_message_size, 1, fp_sei_);
+                    }    
+                    sei_buffer += sei_message[i].sei_message_size;
+                }
+            }
+            free(sei_message_display_q_[pDispInfo->picture_index].sei_data);
+            sei_message_display_q_[pDispInfo->picture_index].sei_data = NULL; // to avoid double free
+            free(sei_message_display_q_[pDispInfo->picture_index].sei_message);
+            sei_message_display_q_[pDispInfo->picture_index].sei_message = NULL; // to avoid double free
+        }
+    }
+    // vp_frames_ffmpeg_.size() is empty, wait for decoding to finish
+    // this will happen during PopFrame()
+    AVFrame *p_av_frame;
+    //std::cout << "Before pop frame " << std::endl;
+#if NO_DECODE_THREAD
```

Comment:
As mentioned above, it is recommended to use an environment variable to check if threading is disabled.

Reply:
Please make sure all AVC/HEVC/AV1 conformance streams pass with the FFMPEG path before merging. The conformance streams test various corner cases related to resolution change, frame flushing or skipping, frame or field (AVC case), etc., which can be affected by the utility code above FFMPEG AV lib.

Reply:
@Jeff. This is mainly for cutomer to use FFMpeg decoding option for HEVC. So currently I have only enabled it for that codec. Once we merge to tot develop, we can add all codec support. We will need to add those as future PRs